ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/Genesis.hs

```diff
@@ -0,0 +1,601 @@
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+
+-- | Genesis decision logic
+--
+-- This module contains the part of the block fetch decisions process that is
+-- specific to the bulk sync mode. This logic reuses parts of the logic for the
+-- deadline mode, but it is inherently different.
+--
+-- Definitions:
+--
+-- - Let @inflight :: peer -> Set blk@ be the outstanding blocks, those that
+--   have been requested and are expected to arrive but have not yet.
+--
+-- - Let @peersOrder@ be an order of preference among the peers. This order is
+--   not set in stone and will evolve as we go.
+--
+-- - Let @currentPeer :: Maybe peer@ be the “current peer” with which we are
+--   interacting. If it exists, this peer must be the best according to
+--   @peersOrder@, and the last fetch request must have been sent to them.
+--
+-- - Let @currentStart :: Time@ be the latest time a fetch request was sent
+--   while there were no outstanding blocks.
```

Comment:
Being caught up can also be a reason for having no outstanding/in-flight blocks; but it isn't the only case. The more usual case would be that we just started to download blocks from a new peer (because of disconnection, or rotation of the dynamo); this is what we want to detect here.